okhttp/src/main/java/okhttp3/internal/tls/OkHostnameVerifier.kt

```diff
@@ -14,130 +14,79 @@
-    for (String altName : altNames) {
-      if (verifyHostname(hostname, altName)) {
-        return true;
-      }
+  /** Returns true if `certificate` matches `ipAddress`.  */
+  private fun verifyIpAddress(ipAddress: String, certificate: X509Certificate): Boolean {
+    return getSubjectAltNames(certificate, ALT_IPA_NAME).any {
+      ipAddress.equals(it, ignoreCase = true)
     }
-    return false;
-  }
-
-  public static List<String> allSubjectAltNames(X509Certificate certificate) {
-    List<String> altIpaNames = getSubjectAltNames(certificate, ALT_IPA_NAME);
-    List<String> altDnsNames = getSubjectAltNames(certificate, ALT_DNS_NAME);
-    List<String> result = new ArrayList<>(altIpaNames.size() + altDnsNames.size());
-    result.addAll(altIpaNames);
-    result.addAll(altDnsNames);
-    return result;
   }
 
-  private static List<String> getSubjectAltNames(X509Certificate certificate, int type) {
-    List<String> result = new ArrayList<>();
-    try {
-      Collection<?> subjectAltNames = certificate.getSubjectAlternativeNames();
-      if (subjectAltNames == null) {
-        return Collections.emptyList();
-      }
-      for (Object subjectAltName : subjectAltNames) {
-        List<?> entry = (List<?>) subjectAltName;
-        if (entry == null || entry.size() < 2) {
-          continue;
-        }
-        Integer altNameType = (Integer) entry.get(0);
-        if (altNameType == null) {
-          continue;
-        }
-        if (altNameType == type) {
-          String altName = (String) entry.get(1);
-          if (altName != null) {
-            result.add(altName);
-          }
-        }
-      }
-      return result;
-    } catch (CertificateParsingException e) {
-      return Collections.emptyList();
+  /** Returns true if `certificate` matches `hostname`.  */
+  private fun verifyHostname(hostname: String, certificate: X509Certificate): Boolean {
+    val hostname = hostname.toLowerCase(Locale.US)
+    return getSubjectAltNames(certificate, ALT_DNS_NAME).any {
+      verifyHostname(hostname, it)
     }
   }
 
   /**
-   * Returns {@code true} iff {@code hostname} matches the domain name {@code pattern}.
+   * Returns true if [hostname] matches the domain name [pattern].
    *
    * @param hostname lower-case host name.
-   * @param pattern domain name pattern from certificate. May be a wildcard pattern such as {@code
-   * *.android.com}.
+   * @param pattern domain name pattern from certificate. May be a wildcard pattern such as
+   *     `*.android.com`.
    */
-  public boolean verifyHostname(String hostname, String pattern) {
+  fun verifyHostname(hostname: String?, pattern: String?): Boolean {
+    var hostname = hostname
+    var pattern = pattern
     // Basic sanity checks
-    // Check length == 0 instead of .isEmpty() to support Java 5.
-    if ((hostname == null) || (hostname.length() == 0) || (hostname.startsWith("."))
-        || (hostname.endsWith(".."))) {
+    if (hostname == null ||
```

Comment:
isNullOrEmpty()

Reply:
Fixed. This is better if we have contracts (we do) and worse if we didn’t.